### DISCUSSION TITLE
WalletConnect methods

### DISCUSSION THREAD

**Discussion by chezhe** [near/wallet-selector#472] [General]
I saw WalletConnect in wallet-selector allowed methods changed from 6.0.x to 7.0.x.
Currently, it's changed into just get `signedTransaction` with the given public key. When I integrate it, I met a problem.
`KeyPair` not support sign with a given public key https://github.com/near/near-api-js/blob/7f16b10ece3c900aebcedf6ebc660cc9e604a242/packages/near-api-js/src/utils/key_pair.ts#L135.

**Comment by kujtimprenkuSQA** [comment]
Hey, @chezhe thanks for raising this issue.

It's true that the name of the methods has been changed from 7.x this has been done to follow the NEP which was approved and merged: https://github.com/near/NEPs/pull/368
https://github.com/near/NEPs/blob/master/specs/Standards/Wallets/BridgeWallets.md

Although I am not sure what you've tried I can confirm that WalletConnect the `wallet-connect` package works correctly at the moment.

To test it out 
- Clone our repo
- Build all the packages 
- Serve one of the examples
- Clone the WalletConnect web-examples from here: https://github.com/amirsaran3/web-examples/tree/SQC-286-update-wallet-connect-web-examples This branch is up to date with the latest changes of the web-examples so when you clone it make sure to checkout to the **SQC-286-update-wallet-connect-web-examples** branch. To run these web examples there is detailed info in the NEP PR above.

Please also have a look at the source code of the `wallet-connect` package you might find some useful info on how we create and signd transactions there: https://github.com/near/wallet-selector/tree/main/packages/wallet-connect

**Reply by chezhe** [reply]
Hi, thanks for reply.
I'm build a mobile wallet which supports WalletConnect 2.
I used to integrate wallet-selector 6.0.x, and it works.
Currently, I'm integrating 7.0.3. I have make `near_signIn`, `near_signOut` work.
But `near_signTransaction` is a little confusing, because it send the raw transaction to wallet client, I tried sign the transaction with the wallet private key and return the signed transaction. It turns out `Transaction is not signed with the given public key`.
As I understand, the transaction is signed by private key rather than public key, also `KeyPair` in `near-api-js` is initialized by secret key(same as private key).

**Reply by chezhe** [reply]
`KeyPair` constructor code as below:
```
 constructor(secretKey) {
        super();
        const keyPair = tweetnacl_1.default.sign.keyPair.fromSecretKey((0, serialize_1.base_decode)(secretKey));
        this.publicKey = new PublicKey({ keyType: KeyType.ED25519, data: keyPair.publicKey });
        this.secretKey = secretKey;
    }
```
It generates a new public key rather than the public key from https://github.com/near/wallet-selector/blob/4128764132e542358bc2458c959018647da23d83/packages/wallet-connect/src/lib/wallet-connect.ts#L349, then how can I sign the transaction with this given public key?

**Comment by kujtimprenkuSQA** [comment]
Maybe you can find this helpful not really sure if it will clear the issue for you when `near_signTransaction` is triggered this is the code that's being run in the wallet in `web-examples` above:
https://github.com/amirsaran3/web-examples/blob/SQC-286-update-wallet-connect-web-examples/wallets/react-wallet-v2/src/utils/NearRequestHandlerUtil.ts#L58-L71

Which eventually calls the `signTransactions` here:
https://github.com/amirsaran3/web-examples/blob/003b8b535dc77ab12e8db2d952bf4d903aed0bf6/wallets/react-wallet-v2/src/lib/NearLib.ts#L281-L308

Also, check the static `init` method in the link above in the NearWallet class.

**Reply by chezhe** [reply]
I find the problem. Before the wallet-connect call sign transaction, it calls `getAccounts`. At first, I respond with the account from `near_signIn`, then I get the error message above.
I clone the repo https://github.com/amirsaran3/web-examples , I find it responds `getAccounts` with the fullAccess public key and use it to sign transaction rather than the `near_signIn` one.

**Comment by max-mainnet** [comment]
Hi @chezhe ,this is Max from Ref Finance.  So glad to hear that you are building mobile wallet to connect to wallet-selector with walletConnect2.0. I am trying to find any wallets I could test wallet-selector + walletConnect2.0 + Fireblocks, so far as I know, no official wallet support this condition right ? Have you heard anything about Fireblocks? Thanks for your time !

**Reply by chezhe** [reply]
I know a little about Fireblocks. My wallet will be out soon, I'm integrating wallet-selector 7.0.x, because they removed near_signAndSendTransaction(s) methods, I have to adapt it.

**Reply by chezhe** [reply]
Hi, Max, I have a issue to talk to you which is about the ref.finance on our mobile wallet, do you mind I send you an email?

**Reply by max-mainnet** [reply]
max@ref.finance, anytime
